### DISCUSSION TITLE
How to setup a local workspace to start checking the code and build it to test

### DISCUSSION THREAD

**Discussion by jlmanohar** [nextcloud/recognize#656] [Q&A] [answered]
Hi, 

I am a java developer by profession but a complete newbie when it comes to javascript, how do i setup the local environment so that i can build and test any changes i make ? I would like to work on this project but need help in setting up my environment and build process

I dont have any IDE preference, but have vs code already installed, along with php 8.1.2, node 12.22.9, node 8.5.1

**Comment by phil-lipp** [comment]
I would also appreciate a (short) explanation on how to do this. Pointing at the right resources to look at would already help a lot.
This would allow me and others to test ideas, play around or beta test changes and thus enable a more involved development process for the Nextcloud/recognize community.

**Comment by marcelklehr** [comment, marked as answer]
Really cool! I'm glad you're interested to dabble with this repo!

You'll need

* php >=7.4 (Nextcloud will require 8.0 starting with v26)
* node >=14
* npm >=8
* composer >= 2

To get started:

1. Clone this repo
2. Run `npm install` to install js dependencies
3. Run `composer install` to install php dependencies
4. `npm run build` (or during development `npm run watch`) builds the settings UI based on vue.js
5. You can manually run the classifier scripts using `node src/classifier_*.js <filepath>`
6. If you put the whole repo at nextcloud/apps/recognize, you can enable the app in the UI and run it in nextcloud

Feel free to ask questions here in the forum or in the matrix chatroom whenever you're stuck. I may take some time to respond but I intend to respond always.

**Reply by jlmanohar** [reply]
Thanks, let me try that, i'm getting some warnings , i installed latest lts of node 18.x, some of the module says required 16. But its just a warning, build seems to be a success with 3 warnings shown.

**Reply by jlmanohar** [reply]
Also to anyone having issues with composer, I would recommend running composer update before doing composer install, also install php-xml/php8.1-xml, and php8.1-curl/php-curl (php or php8.1, whichever one is relevant to your distro) as without that composer seems to fail.

**Comment by jlmanohar** [comment]
Is there a way i can overlay the outpt json onto the src image to see what it recognized ? Also it says that error with input image format but it does work fine

`node src/classifier_*.js /home/jlmanohar/Downloads/54129-DSC06877.png`

**Output:**

2023-02-09 21:39:02.538637: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations: AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Error: Expected image (BMP, JPEG, PNG, or GIF), but got unsupported image type
at getImageType (/home/jlmanohar/Documents/code/nextcloud/recognize/node_modules/@tensorflow/tfjs-node/dist/image.js:353:15)
at Object.decodeImage (/home/jlmanohar/Documents/code/nextcloud/recognize/node_modules/@tensorflow/tfjs-node/dist/image.js:198:21)
at main (/home/jlmanohar/Documents/code/nextcloud/recognize/src/classifier_faces.js:56:28)
[]
Error: Expected image (BMP, JPEG, PNG, or GIF), but got unsupported image type
at getImageType (/home/jlmanohar/Documents/code/nextcloud/recognize/node_modules/@tensorflow/tfjs-node/dist/image.js:353:15)
at Object.decodeImage (/home/jlmanohar/Documents/code/nextcloud/recognize/node_modules/@tensorflow/tfjs-node/dist/image.js:198:21)
at main (/home/jlmanohar/Documents/code/nextcloud/recognize/src/classifier_faces.js:56:28)
[]
Error: Expected image (BMP, JPEG, PNG, or GIF), but got unsupported image type
at getImageType (/home/jlmanohar/Documents/code/nextcloud/recognize/node_modules/@tensorflow/tfjs-node/dist/image.js:353:15)
at Object.decodeImage (/home/jlmanohar/Documents/code/nextcloud/recognize/node_modules/@tensorflow/tfjs-node/dist/image.js:198:21)
at main (/home/jlmanohar/Documents/code/nextcloud/recognize/src/classifier_faces.js:56:28)
[]
Error: Expected image (BMP, JPEG, PNG, or GIF), but got unsupported image type
at getImageType (/home/jlmanohar/Documents/code/nextcloud/recognize/node_modules/@tensorflow/tfjs-node/dist/image.js:353:15)
at Object.decodeImage (/home/jlmanohar/Documents/code/nextcloud/recognize/node_modules/@tensorflow/tfjs-node/dist/image.js:198:21)
at main (/home/jlmanohar/Documents/code/nextcloud/recognize/src/classifier_faces.js:56:28)
[]
Error: Expected image (BMP, JPEG, PNG, or GIF), but got unsupported image type
at getImageType (/home/jlmanohar/Documents/code/nextcloud/recognize/node_modules/@tensorflow/tfjs-node/dist/image.js:353:15)
at Object.decodeImage (/home/jlmanohar/Documents/code/nextcloud/recognize/node_modules/@tensorflow/tfjs-node/dist/image.js:198:21)
at main (/home/jlmanohar/Documents/code/nextcloud/recognize/src/classifier_faces.js:56:28)
[]
[{"angle":{"roll":19,"pitch":-4,"yaw":-25},"vector":{"0":-0.17705905437469482,"1":0.041112933307886124,"2":0.07457832992076874,"3":-0.0053581614047288895,"4":-0.07995298504829407,"5":-0.13841553032398224,"6":-0.04642868787050247,"7":-0.11967705190181732,"8":0.16007553040981293,"9":-0.07566443085670471,"10":0.23981228470802307,"11":-0.013764716684818268,"12":-0.14470435678958893,"13":-0.06445036083459854,"14":-0.04613841325044632,"15":0.09858343750238419,"16":-0.1490897834300995,"17":-0.1387210190296173,"18":-0.08404971659183502,"19":-0.06529597193002701,"20":-0.03168449178338051,"21":-0.07849584519863129,"22":0.043784864246845245,"23":0.1301310956478119,"24":-0.11958047747612,"25":-0.3501020669937134,"26":-0.11806699633598328,"27":-0.14622747898101807,"28":0.12641116976737976,"29":-0.019322138279676437,"30":-0.05103977024555206,"31":0.06593643873929977,"32":-0.17341654002666473,"33":-0.07433050125837326,"34":0.03364265710115433,"35":0.08768776804208755,"36":0.010196711868047714,"37":-0.04382152855396271,"38":0.24146033823490143,"39":-0.015799004584550858,"40":-0.17849677801132202,"41":-0.04075472429394722,"42":0.019026439636945724,"43":0.2680184841156006,"44":0.17661693692207336,"45":0.025239529088139534,"46":0.07607395946979523,"47":-0.016912031918764114,"48":0.09521090984344482,"49":-0.22082844376564026,"50":0.04057078808546066,"51":0.14864496886730194,"52":0.15544451773166656,"53":-0.013675481081008911,"54":0.1123073399066925,"55":-0.18316243588924408,"56":0.026798464357852936,"57":0.07468937337398529,"58":-0.1962452232837677,"59":0.0872812271118164,"60":0.0002223961055278778,"61":-0.06460420787334442,"62":-0.058584995567798615,"63":-0.03633447736501694,"64":0.2500717341899872,"65":0.1784338653087616,"66":-0.10456956177949905,"67":-0.11836837232112885,"68":0.23726557195186615,"69":-0.19520048797130585,"70":0.03339575603604317,"71":0.09572840481996536,"72":-0.10903645306825638,"73":-0.1649564504623413,"74":-0.28856462240219116,"75":-0.002640375867486,"76":0.496954083442688,"77":0.1827596127986908,"78":-0.158323734998703,"79":0.0655197724699974,"80":-0.12529873847961426,"81":-0.05518524348735809,"82":0.009756422601640224,"83":0.05384785681962967,"84":-0.1859404742717743,"85":0.06597758829593658,"86":-0.1091633290052414,"87":0.005664519965648651,"88":0.15931706130504608,"89":0.07267804443836212,"90":-0.0301506407558918,"91":0.16381831467151642,"92":-0.005311794579029083,"93":0.036079198122024536,"94":0.0558207631111145,"95":-0.03682442754507065,"96":-0.10515057295560837,"97":-0.0897979587316513,"98":-0.11498995125293732,"99":-0.03140924125909805,"100":0.009581334888935089,"101":-0.0762287825345993,"102":0.03529263287782669,"103":0.07054665684700012,"104":-0.19608443975448608,"105":0.05330269783735275,"106":-0.015558943152427673,"107":-0.03710074722766876,"108":-0.04641089588403702,"109":0.10469256341457367,"110":-0.12025843560695648,"111":-0.055893853306770325,"112":0.14862023293972015,"113":-0.2539002597332001,"114":0.16228307783603668,"115":0.16484375298023224,"116":0.05692610144615173,"117":0.14628304541110992,"118":0.08649072796106339,"119":0.009001409634947777,"120":-0.0076965149492025375,"121":-0.017502278089523315,"122":-0.0592280812561512,"123":-0.059408821165561676,"124":0.0034671640023589134,"125":-0.11458522826433182,"126":0.09704864025115967,"127":0.03497621789574623},"x":0.28315505385398865,"y":0.29590031688164403,"height":0.31324128181703625,"width":0.18480783700942993,"score":0.9719710350036621},{"angle":{"roll":18,"pitch":-14,"yaw":186},"vector":{"0":-0.11856856197118759,"1":0.09666471183300018,"2":0.07477124035358429,"3":-0.070732980966568,"4":-0.08683697134256363,"5":-0.008271489292383194,"6":-0.04258282855153084,"7":-0.1300143301486969,"8":0.19298212230205536,"9":-0.1453896015882492,"10":0.19893217086791992,"11":0.004225276410579681,"12":-0.27959710359573364,"13":-0.016658732667565346,"14":-0.04851536080241203,"15":0.10108073055744171,"16":-0.07149435579776764,"17":-0.1024329662322998,"18":-0.09939558058977127,"19":-0.08365113288164139,"20":0.05141439288854599,"21":-0.01922806352376938,"22":0.07839079201221466,"23":0.04717354103922844,"24":-0.1847432255744934,"25":-0.329873651266098,"26":-0.11249097436666489,"27":-0.10425560176372528,"28":0.017138861119747162,"29":-0.07764407247304916,"30":-0.08437799662351608,"31":0.0034857764840126038,"32":-0.23890727758407593,"33":-0.07003726810216904,"34":0.0011156592518091202,"35":0.11605405807495117,"36":-0.05027315765619278,"37":-0.03996259346604347,"38":0.10761511325836182,"39":0.017438840121030807,"40":-0.18449479341506958,"41":-0.06692256033420563,"42":0.03359586000442505,"43":0.20969264209270477,"44":0.18036802113056183,"45":0.08379264920949936,"46":0.08291522413492203,"47":-0.13199564814567566,"48":0.058730050921440125,"49":-0.21156983077526093,"50":0.08191473037004471,"51":0.10338470339775085,"52":0.1469087302684784,"53":-0.01211024820804596,"54":0.1222696304321289,"55":-0.1860867589712143,"56":0.04388891160488129,"57":0.10365376621484756,"58":-0.21894802153110504,"59":0.0562160424888134,"60":0.0781228095293045,"61":-0.020870624110102654,"62":-0.07428935170173645,"63":-0.021277382969856262,"64":0.2555937170982361,"65":0.1409476101398468,"66":-0.1449059545993805,"67":-0.0887167677283287,"68":0.23104673624038696,"69":-0.18105655908584595,"70":0.03618396818637848,"71":0.08042330294847488,"72":-0.1042139083147049,"73":-0.22728431224822998,"74":-0.31605076789855957,"75":-0.007946470752358437,"76":0.4909053146839142,"77":0.1996738314628601,"78":-0.1083146333694458,"79":0.0475768968462944,"80":-0.1540096551179886,"81":0.011825177818536758,"82":0.042025621980428696,"83":0.05380711704492569,"84":-0.15199047327041626,"85":0.06825195997953415,"86":-0.06333351880311966,"87":0.076260507106781,"88":0.1703282594680786,"89":-0.0076479800045490265,"90":-0.040540505200624466,"91":0.1704091727733612,"92":-0.016139954328536987,"93":-0.00414658896625042,"94":0.014809895306825638,"95":0.020100921392440796,"96":-0.17700296640396118,"97":-0.05164347216486931,"98":-0.13294248282909393,"99":-0.04221539944410324,"100":-0.03909709304571152,"101":-0.09531287103891373,"102":0.01081334799528122,"103":0.10327610373497009,"104":-0.22745412588119507,"105":0.0813460499048233,"106":-0.017265960574150085,"107":-0.10532134771347046,"108":-0.03247801214456558,"109":0.002791881561279297,"110":-0.16234679520130157,"111":-0.06635408103466034,"112":0.13068023324012756,"113":-0.3299696445465088,"114":0.11970173567533493,"115":0.20575666427612305,"116":0.0214658435434103,"117":0.15988677740097046,"118":0.11751475930213928,"119":0.04890081286430359,"120":0.05716770887374878,"121":-0.032268375158309937,"122":-0.14904017746448517,"123":-0.07326261699199677,"124":0.056902673095464706,"125":-0.10635127127170563,"126":0.11265557259321213,"127":0.0014145136810839176},"x":0.4546951651573181,"y":0.15080073292304344,"height":0.37889448149113364,"width":0.2096097469329834,"score":0.9488032460212708}]
